src/java.base/share/classes/java/util/zip/GZIPInputStream.java

```diff
@@ -76,10 +88,75 @@ private void ensureOpen() throws IOException {
+     * allows concatenated compressed streams and ignores extra bytes.
+     *
+     * @param in the input stream
+     *
+     * @throws    ZipException if a GZIP format error has occurred or the
+     *                         compression method used is unsupported
+     * @throws    NullPointerException if {@code in} is null
+     * @throws    IOException if an I/O error has occurred
+     */
+    public GZIPInputStream(InputStream in) throws IOException {
+        this(in, 512, true, true);
+    }
+
+    /**
+     * Creates a new input stream with the specified buffer size,
+     * support for concatenated streams, and tolerance for extra bytes.
+     *
+     * <p>
+     * When {@code allowConcatenation} is true, this class will attempt to decode any data that
+     * follows a GZIP trailer frame as the GZIP header frame of a new compressed data stream
+     * and proceed to decompress it. As a result, arbitrarily many consecutive compressed data
+     * streams in the underlying input will be read back as a single uncompressed stream. When
+     * {@code allowConcatenation} is false, decompression stops after the first GZIP trailer
+     * frame encountered, and any additional bytes are considered extraneous.
+     *
+     * <p>
+     * The {@code ignoreExtraBytes} flag controls the behavior when extraneous data appears,
+     * i.e., when any data follows the GZIP trailer frame (if {@code allowConcatenation} is
+     * false), or when any data other than a valid GZIP header frame follows a GZIP trailer
+     * frame (if {@code allowConcatenation} is true). When {@code ignoreExtraBytes} is true,
+     * the unexpected data, and any {@link IOException} thrown while trying to read that
+     * data, are simply discarded, and EOF is returned; when {@code ignoreExtraBytes} is
+     * false, any unexpected data triggers an {@link IOException}, and any {@link IOException}
+     * thrown while trying to read a concatenated GZIP header frame is propagated to the caller.
+     * In other words, every byte of the underlying data stream is part of a complete and valid
+     * compressed data stream, or else an {@link IOException} is guaranteed to be thrown.
+     *
+     * @apiNote The original behavior of this class is replicated by setting both
+     * {@code allowConcatenation} and {@code ignoreExtraBytes} to true. However,
+     * enabling {@code ignoreExtraBytes} is discouraged because of its imprecision in
+     * how many additional bytes are read and the possibility that {@link IOException}s
+     * and data corruption in the underlying stream can go unnoticed.
+     *
+     * @param in the input stream
+     * @param size the input buffer size
+     * @param allowConcatenation true to allow multiple concatenated compressed data streams,
+     *                           or false to expect exactly one compressed data stream
+     * @param ignoreExtraBytes true to tolerate and ignore extra bytes, false to throw
```

Comment:
The term "extra" here feels somewhat open to interpretation. Specifically, "extra" sounds like something that is out of the ordinary, but not uncommon or wrong. It could be used when describing an optional feature in a format specification.

The API description referes to "unexpected data". Perhaps the word "unexpected" is more precise term to use in this option? So `ignoreUnexpectedBytes` or `ignoreUnexpectedData`.

I think my eyebrows would be raised more when seeing someone ignoring 'unexpected data' rather than when ignoring 'extra data'.

I know this might smell of bikeshedding, but naming is important (and hard!).

Reply:
> The term "extra" here feels somewhat open to interpretation. Specifically, "extra" sounds like something that is out of the ordinary, but not uncommon or wrong. It could be used when describing an optional feature in a format specification.
> 
> The API description referes to "unexpected data". Perhaps the word "unexpected" is a more precise and descriptive term to use in this option? So ignoreUnexpectedBytes or ignoreUnexpectedData.

This is a good point.. because it's actually a bit subtle what kind of data is being referred to here as being "ignored". Even "unexpected" doesn't quite capture it.

To be precise, here's what bytes the  `ignoreExtraBytes` parameter is going to cause to be ignored:
* In the case `allowConcatenation == false`, it really does refer to "extra" or "extraneous" data, that is: one or more bytes having any value appearing _after_ the end of the GZIP trailer frame.
* In the case `allowConcatenation == true`, it means a truncated or invalid GZIP _header frame_ appearing after the end of any GZIP trailer frame.

The second case means "unexpected" seems wrong because why would unexpected data in a header frame be treated any differently from unexpected data any other frame (which of course is not going to be ignored but instead will trigger an exception)?

So maybe this is just more evidence that we shouldn't use boolean parameters - because they're not really orthogonal.

What about something like:

```java
public enum ConcatPolicy {
    DISALLOW_STRICT,    // one GZIP stream, nothing else
    DISALLOW_LENIENT,   // one GZIP stream, ignore any extra byte(s)
    ALLOW_STRICT,       // N GZIP streams, nothing else
    ALLOW_LENIENT;      // N GZIP streams, stop at, and ignore, an invalid header frame
}
```
The legacy behavior would of course correspond to `ALLOW_LENIENT`.

Reply:
Another thought:

Are we sure we would want to introduce a new mode now which does _not_ allow concatenation, but _does_ ignore trailing data?

If the ignore mode is really discouraged, why open a new mode of config allowing it?

In other words, could we instead (at least conceptually)  have the policies LEGACY, SINGLE_STREAM, CONCATENATED_STREAM where the latter two always reject trailing data?

Reply:
> Are we sure we would want to introduce a new mode now which does not allow concatenation, but does ignore trailing data?

I don't see it as necessary. It's certainly not a case that anyone is currently relying on. So it's fine with me to omit it.

If are calling this a "concatenated stream policy" then we could have:
```java
public enum ConcatPolicy {
    DISALLOW,
    ALLOW,
    ALLOW_LENIENT;
}
```
(I'm avoiding `LEGACY` which makes sense today but might make less sense N years from now.)